apis/apps/v1alpha1/opsrequest_webhook.go

```diff
@@ -416,12 +416,134 @@ func (r *OpsRequest) validateHorizontalScaling(_ context.Context, _ client.Clien
+	offlineOrOnlineInsCountMap := map[string]int32{}
+	for _, insName := range hScaleInstanceNames {
+		insTplName := GetInstanceTemplateName(clusterName, componentName, insName)
+		offlineOrOnlineInsCountMap[insTplName]++
+	}
+	return offlineOrOnlineInsCountMap
+}
+
+func (r *OpsRequest) validateHorizontalScalingSpec(hScale HorizontalScaling, compSpec ClusterComponentSpec, clusterName string, isSharding bool) error {
+	scaleIn := hScale.ScaleIn
+	scaleOut := hScale.ScaleOut
+	if hScale.Replicas != nil && (scaleIn != nil || scaleOut != nil) {
+		return fmt.Errorf(`"replicas" has been deprecated and cannot be used with "scaleOut" and "scaleIn"`)
+	}
+	if hScale.Replicas != nil {
+		return nil
+	}
+	if lastCompConfiguration, ok := r.Status.LastConfiguration.Components[hScale.ComponentName]; ok {
+		// use last component configuration snapshot
+		compSpec.Instances = lastCompConfiguration.Instances
+		compSpec.Replicas = *lastCompConfiguration.Replicas
+		compSpec.OfflineInstances = lastCompConfiguration.OfflineInstances
+	}
+	compInsTplMap := map[string]int32{}
+	for _, v := range compSpec.Instances {
+		compInsTplMap[v.Name] = v.GetReplicas()
+	}
+	// Rules:
+	// 1. length of offlineInstancesToOnline or onlineInstancesToOffline can't greater than the configured replicaChanges for the component.
+	// 2. replicaChanges for component must greater than or equal to the sum of replicaChanges configured in instance templates.
+	validateHScaleOperation := func(replicaChanger ReplicaChanger, newInstances []InstanceTemplate, offlineOrOnlineInsNames []string, isScaleIn bool) error {
+		msgPrefix := "ScaleIn:"
+		hScaleInstanceFieldName := "onlineInstancesToOffline"
+		if !isScaleIn {
+			msgPrefix = "ScaleOut"
+			hScaleInstanceFieldName = "offlineInstancesToOnline"
+		}
+		if !isSharding && replicaChanger.ReplicaChanges != nil && len(offlineOrOnlineInsNames) > int(*replicaChanger.ReplicaChanges) {
+			return fmt.Errorf(`the length of %s can't be greater than the "replicaChanges" for the component`, hScaleInstanceFieldName)
+		}
+		offlineOrOnlineInsCountMap := r.CountOfflineOrOnlineInstances(clusterName, hScale.ComponentName, offlineOrOnlineInsNames)
+		insTplChangeMap := map[string]int32{}
+		allReplicaChanges := int32(0)
+		for _, v := range replicaChanger.Instances {
+			compInsReplicas, ok := compInsTplMap[v.Name]
+			if !ok {
+				return fmt.Errorf(`%s cannot find the instance template "%s" in component "%s"`,
+					msgPrefix, v.Name, hScale.ComponentName)
+			}
+			if isScaleIn && v.ReplicaChanges > compInsReplicas {
+				return fmt.Errorf(`%s "replicaChanges" of instanceTemplate "%s" can't be greater than %d`,
+					msgPrefix, v.Name, compInsReplicas)
+			}
+			allReplicaChanges += v.ReplicaChanges
+			insTplChangeMap[v.Name] = v.ReplicaChanges
+		}
+		for insTplName, replicaCount := range offlineOrOnlineInsCountMap {
+			replicaChanges, ok := insTplChangeMap[insTplName]
+			if !ok {
+				allReplicaChanges += replicaCount
```

Comment:
if ok, should also execute allReplicaChanges += replicaCount

Reply:
if ok, means the replicaChanges has been specified in` instances.replicaChanges` and will add it.